### PR TITLE
ssh_student for Docker

### PR DESCRIPTION
**Pull request by darkyne**
This adds the ssh_student functionality for docker environment (not working with Kata).
Basically, when the grading_script hits the ssh_student function, it starts a student_container which will start a ssh server and send to the front-end the commands for the student to connect into the student_container. When the student leaves the ssh session, the grading_script continues to run.

A very simple example is the following:
image

Notes:
- If no one connects to the ssh server, it times out after 2min and the grading script continues.
- The files modified in the student directory can be used for evaluation in the grading container

Still to improve: (marked with "TODO" in the files)
- If no one connects, it might be better to send an error message (and not to continue the script)
- If the teacher configures the task with a very small time-out (for example less than 5 minutes), it might be useful to send a warning message to tell the teacher to put a longer time limit.
- The student-container (when launched via ssh_student) is using "bridge" as network mode. Maybe it is not the best configuration.
- There are some duplicated code between INGInious and _run_student_intern (setLimits, execute_process, launch sshd and wait for user). It should be possible to put this in a file (example utils.py) and to import it instead of duplications. I had a quick try but didn't work.
- Some communication between parts are sometimes not perfectly designed (but working), example [this](https://github.com/darkyne/INGInious/blob/aac99d98a255e04ae508b897bde2b565cc1508de/base-containers/base/inginious_container_api/run_student.py#L116).
-  Block this functionality if the teacher did not hit "allow internet access" in the task configuration. Indeed, the student_container to allow ssh fonctionalities would require a bridge network mode.
- Verify the behaviour if a student connects with ssh and kills the sshd process.

Technical explanations:
- "ssh_student" calls "run_student" with the parameter "ssh" set to True.
- The student_container is launched with _run_student_intern and receives a message which was sent [here](https://github.com/darkyne/INGInious/blob/aac99d98a255e04ae508b897bde2b565cc1508de/base-containers/base/inginious_container_api/run_student.py#L105). This was originaly the way how run_student sent the cmd to the student_container. Now it also sends the "ssh" boolean.
- When receiving this message, the student_container starts a ssh server only if the "ssh" boolean is True [(here)](https://github.com/darkyne/INGInious/blob/aac99d98a255e04ae508b897bde2b565cc1508de/base-containers/base/bin/_run_student_intern#L115). Otherwise, it runs the cmd as usually.
- After the student_container started the ssh server, it sends a message [(here)](https://github.com/darkyne/INGInious/blob/aac99d98a255e04ae508b897bde2b565cc1508de/base-containers/base/bin/_run_student_intern#L137) with id and password.
- This message is received by the grading_container [(here)](https://github.com/darkyne/INGInious/blob/aac99d98a255e04ae508b897bde2b565cc1508de/base-containers/base/inginious_container_api/run_student.py#L116) and sent to the agent [(here)](https://github.com/darkyne/INGInious/blob/aac99d98a255e04ae508b897bde2b565cc1508de/base-containers/base/inginious_container_api/run_student.py#L122).
- The agent received this message [(here)](https://github.com/darkyne/INGInious/blob/aac99d98a255e04ae508b897bde2b565cc1508de/inginious/agent/docker_agent/__init__.py#L576). It adds the correct external port to reach the concerned student_container (based on the container_id) and sends this to the front-end allowing the student to connect to the student_container.

Probably many little improvements to do but I hope no huge issue.

### CODE REVIEW

File: base-containers/base/inginious_container_api/run_student.py
Comment (GuillaumeDerval) [inline base-containers/base/inginious_container_api/run_student.py:116]: You have to ensure that the message length is < 250 bytes

File: base-containers/base/bin/_run_student_intern
Comment (GuillaumeDerval) [inline base-containers/base/bin/_run_student_intern:115]: Is it possible for this to be in its own command? It's strange to run `run_student` without cmd and for it to start the ssh server itself.

Three reasons:
1) I think the same could be achieved with `run_student start_the_ssh_server_please`
2) The behavior of run_student would be more consistent
3) It would be nice to parametrize more the way the ssh server works. For exemple, here we always start `bash` when someone connects, but we may want to start other things: `run_student start_the_ssh_server_please --with-shell=zsh`

File: inginious/agent/docker_agent/__init__.py
Comment (darkyne) [inline inginious/agent/docker_agent/__init__.py:553]: Woops. I meant "frontend" not "backend"

File: base-containers/base/bin/_run_student_intern
Comment (GuillaumeDerval) [reply base-containers/base/bin/_run_student_intern:115]: I see a problem with this: it doesn't warn the agent that the student container needs a ssh port. We can keep the ssh flag for `run_student`: `run_student --port=22 start_the_ssh_server_please --with-shell=zsh` or something like that.

This would make the system more versatile too.

File: base-containers/base/inginious_container_api/run_student.py
Comment (darkyne) [reply base-containers/base/inginious_container_api/run_student.py:116]: Now:
The message sent is: message = {"type": "ssh_student", "ssh_user": ssh_user_value, "password": password}
It has a constant size of 240 bytes (indeed sys.getsizeof(message)==240)
The password size is of constant size and users are stored as 0 for worker, 1 for root.

Thus, now I use: ssh_id = msgpack.loads(connection.recv(240))
I hope it resolves the potential issues here :)

File: base-containers/base/bin/_run_student_intern
Comment (anthonygego) [inline base-containers/base/bin/_run_student_intern:100]: It is highly probable that ``sys.getsizeof`` is architecture-dependent. There may be a more appropriate msgpack API to compute this. I'm not even sure this check would be useful at this stage.

File: base-containers/base/inginious_container_api/run_student.py
Comment (anthonygego) [inline base-containers/base/inginious_container_api/run_student.py:118]: There should be a msgpack API to compute exactually at runtime the expected message size, rather than hardcoding it. This is typically the kind of stuff that the library should abstract.

File: base-containers/base/bin/ssh_student
Comment (anthonygego) [inline base-containers/base/bin/ssh_student:9]: As you have defined `ssh_student`, maybe it would be more coherent to import it here instead of `run_student`.

File: base-containers/base/inginious_container_api/utils.py
Comment (anthonygego) [inline base-containers/base/inginious_container_api/utils.py:58]: `permit_root_login = ...`

**Review by anthonygego — CHANGES_REQUESTED** [review]

File: base-containers/base/bin/_run_student_intern
Comment (darkyne) [reply base-containers/base/bin/_run_student_intern:100]: Indeed, the way I checked its size was incorrect.
I was way better to tansform the message into bytes with msgpack before checking its len.

File: base-containers/base/inginious_container_api/run_student.py
Comment (darkyne) [reply base-containers/base/inginious_container_api/run_student.py:118]: The exact number of bytes can be computed at the student_container side (it's actually a constant: 54 bytes).
But since the whole goal here is to communicate between these 2 sides (student_container and grading_container) I don't see the issue about hardcoding the number of bytes (since it's a consant)  :/

Comment (anthonygego) [reply base-containers/base/inginious_container_api/run_student.py]: It is rarely a good practice to hardcode that kind of thing unless there is an explicit standardization that ensures this will never change. While msgpack should ensure the size of the message is constant through different architectures by its design, you never know if the internal message representation won't change some day as it is not standardized.

These kind of changes actually already happend once when msgpack was upgraded to version 1.0. So you should really rely on a msgpack for this.

The typical way msgpack deal with this is through its [Unpacker class](https://github.com/msgpack/msgpack-python/tree/main#streaming-unpacking). And this is already the way it is implemented in INGInious : https://github.com/UCL-INGI/INGInious/blob/b60326b702022ffe41391e7b7b003b3b3bbb8080/base-containers/base/bin/_run_student_intern#L59

And, while it is commented "It's ugly", there is actually no many other manners to do this with msgpack. It is simply not designed to support non-msgpack messages in the same stream. A possible alternative, in this really specific case, is to pack a dummy message, compute its size at runtime, and this should be the expected value.

It is clear the actual way to solve this "It's ugly" thing is to ensure only msgpack messages are sent in the stream, and I'm not requiring it. I just want to avoid really bad surprises if something changes in msgpack implementation.

Comment (darkyne) [reply base-containers/base/inginious_container_api/run_student.py]: Should be better now, computing the expected size at runtime instead of hardcoding it.

**Comment by darkyne** [comment]
Any news about things still to change here ?
To be honnest, I don't understand why Jenkins tests failed (build step failed) but if anyone can give a clue, it may help :)

**Comment by anthonygego** [comment]
Didn't know you were done. I'll have another check very soon. I think we're still waiting for @GuillaumeDerval concerning the way to unpack msgpack messages. Jenkins fails because we need to update to API key, the current one seems to have expired for security reasons.

**Comment by darkyne** [comment]
Well, when testing everything seems to work correctly and the modifications you requested are done.
About the way the message is unpacked, I went for the byte by byte approach as it was already done previously in the code.

That being said, I'm not especially saying "the work is done" just the modifications requested are done but you might want others things to change, don't hesitate :) I'm working on kata during that time anyway.

File: base-containers/base/bin/_run_student_intern
Comment (GuillaumeDerval) [inline base-containers/base/bin/_run_student_intern:83]: "ron" -> "run"

File: base-containers/base/bin/_run_student_intern
Comment (GuillaumeDerval) [inline base-containers/base/bin/_run_student_intern:101]: The comment is not needed and may become wrong in the future

File: base-containers/base/bin/_run_student_intern
Comment (GuillaumeDerval) [inline base-containers/base/bin/_run_student_intern:100]: Why not using "root" and "worker" instead of 0/1?

File: base-containers/base/bin/ssh_student
Comment (GuillaumeDerval) [inline base-containers/base/bin/ssh_student:11]: Use """ :-)

File: base-containers/base/inginious_container_api/run_student.py
Comment (GuillaumeDerval) [inline base-containers/base/inginious_container_api/run_student.py:25]: Is it possible to have a process that continues to work once the SSH server is started?
(that is, to detach a process that would continue to work in background).

It would be nice to document a small example explaining how to do that.

File: base-containers/base/inginious_container_api/run_student.py
Comment (GuillaumeDerval) [inline base-containers/base/inginious_container_api/run_student.py:45]: Please add a warning about runtimes that do not support root

File: base-containers/base/inginious_container_api/run_student.py
Comment (GuillaumeDerval) [inline base-containers/base/inginious_container_api/run_student.py:103]: "The socket can only work in a ping-pong way, so we need to send a dummy message"

File: base-containers/base/inginious_container_api/ssh_student.py
Comment (GuillaumeDerval) [inline base-containers/base/inginious_container_api/ssh_student.py:11]: Please insert some documentation here

File: base-containers/base/inginious_container_api/utils.py
Comment (GuillaumeDerval) [inline base-containers/base/inginious_container_api/utils.py:13]: I would prefer root="root" and worker="worker". It's easier to read while debugging, it loses some bits while sending packets, but it's not a problem.

File: inginious/agent/docker_agent/__init__.py
Comment (GuillaumeDerval) [inline inginious/agent/docker_agent/__init__.py:586]: Please check that the container_id belongs to one of the student containers of this grading container.

**Review by GuillaumeDerval — CHANGES_REQUESTED** [review]

File: base-containers/base/bin/_run_student_intern
Comment (darkyne) [reply base-containers/base/bin/_run_student_intern:100]: Indeed. Actually encoding the user as a binary value (0 or 1) was to allow a constant size message.
Now since the message size is sent seperately, I will simply use "worker" and "root" instead :)

File: base-containers/base/inginious_container_api/run_student.py
Comment (darkyne) [reply base-containers/base/inginious_container_api/run_student.py:25]: Right now with the curent implementation, the command is started in a thread p and we p.wait() before launching the ssh server...
But actually, it would be easy to add an argument to specify that we want the command to continue in its thread and to stop only when the exercice is finished. Thus in this case, we would not p.wait() directly but instead we would start the ssh server and maybe p.wait or p.kill at the end before exiting the container.

Do you think such an argument would be useful ?
--continue: if you want your command (probably a script) to continue running in background while the student has ssh access.

**Comment by darkyne** [comment]
Just added the option:
--continuous

This allows the command (if there is one) to be run in a background process while the student has ssh access.
When --continuous is not set, the command is completed before starting the ssh server.

As a quick test, I wrote a python script writing something in a file every second and putted the --continuous flag.
When I connected as a student with ssh, I was able to see the file being edited every second.
This background process (running the command) terminates when the student leaves the connection (or when he timesout).

Remark: For the general case of run_student (thus with "ssh" set to False), the new "continuous" argument has no effect (as described it is only useful if "ssh" is also set to True).

I hope this meets the requirements.

**Comment by anthonygego** [comment]
I think that ``--run-in-background``, ``--bg``, or anything that recalls the background process concept would be more appropriate than ``--continuous``. I don't mean to be pedant, but the feature would be more easily understood if it relies on commonly used vocabulary.

**Comment by darkyne** [comment]
Indeed, this terminology is better. Corrected.

File: inginious/agent/docker_agent/__init__.py
Comment (anthonygego) [inline inginious/agent/docker_agent/__init__.py:446]: As this is simple, you can save 3 lines by writing : ``ports_needed = [22] if ssh else []`` :)

**Comment by darkyne** [comment]
Documentation should be better now.
Also did very few little modifications about code-style.
Hope it's fine now :)

File: base-containers/base/inginious_container_api/run_student.py
Comment (anthonygego) [reply base-containers/base/inginious_container_api/run_student.py:126]: This is still to be discussed :-)

Comment (darkyne) [reply base-containers/base/inginious_container_api/run_student.py:126]: True. We can have a word about it during our meeting :)

Comment (darkyne) [reply base-containers/base/inginious_container_api/run_student.py:126]: Added the '!' symbol and noticed no difference when testing, should be fine :)

File: inginious/agent/docker_agent/__init__.py
Comment (anthonygego) [inline inginious/agent/docker_agent/__init__.py:590]: It seems that killing the job here may result in exceptions in ``handle_job_closing``:

    2021-06-30 08:48:54,638 - inginious.agent.docker - ERROR - Exception: ssh for student requires internet access in the task configuration !.
    NoneType: None
    2021-06-30 08:48:54,638 - inginious.backend - INFO - Job ad3d0b58-6368-4a64-97c1-25c14e388a35 finished on agent b'\x00\x19I]\x00'
    2021-06-30 08:48:54,713 - inginious.agent.docker - WARNING - Container 8dce2b6b4432639b526bde7e5acf2c2a378d65b1e9d60e132d64df9dff2d4969 has not given any result
    2021-06-30 08:48:54,723 - inginious.agent.docker - ERROR - Exception in handle_job_closing
    Traceback (most recent call last):
      File "/home/anthony/Documents/inginious-dev/INGInious/inginious/agent/docker_agent/__init__.py", line 781, in handle_job_closing
        await self.send_job_result(info.job_id, result, error_msg, grade, problems, tests, custom, state, archive, stdout, stderr)
      File "/home/anthony/Documents/inginious-dev/INGInious/inginious/agent/__init__.py", line 215, in send_job_result
        raise JobNotRunningException()
    inginious.agent.JobNotRunningException
    2021-06-30 08:48:55,328 - inginious.webapp.requests - INFO - 127.0.0.1:48590 - "- POST /course/test/test_ssh" - 200 OK
    127.0.0.1 - - [30/Jun/2021 08:48:55] "POST /course/test/test_ssh HTTP/1.1" 200 -

Moreover, I don't know where this ``NoneType: None`` comes from in my logs. The error message is not displayed in the webapp. I'm only left with *Internal error* :-(

Comment (darkyne) [reply inginious/agent/docker_agent/__init__.py:590]: Indeed, the error should be correctly handled and the feedback should be more explicit.
The strange NoneType: None origin is surprisingly hard to find :o 
Working on this right now :)

Comment (darkyne) [reply inginious/agent/docker_agent/__init__.py:590]: About the "Nonetype: None" being logged, it's clearly the self._logger.exception("...") that makes this happen...
But no idea why :/

**Comment by anthonygego** [comment]
LGTM. @GuillaumeDerval could you have a quick look at the last commit ?

**Comment by anthonygego** [comment]
@GuillaumeDerval  could you have a quick look at the last commit so that I can merge this part ?

File: base-containers/base/bin/_run_student_intern
Comment (GuillaumeDerval) [inline base-containers/base/bin/_run_student_intern:65]: Maybe add a check here that we are indeed allowed to start "root" in this container.
Is it possible for a compromised grading container to start a root student container on Docker?

Comment (GuillaumeDerval) [reply base-containers/base/bin/_run_student_intern:65]: (it is indeed currently possible since this message comes from the grading container)

File: base-containers/base/inginious_container_api/run_student.py
Comment (GuillaumeDerval) [inline base-containers/base/inginious_container_api/run_student.py:45]: Add a warning saying that it depends on the runtime?

File: inginious/agent/docker_agent/__init__.py
Comment (GuillaumeDerval) [inline inginious/agent/docker_agent/__init__.py:451]: Caution: you should ensure that the ports that were already in `ports` are going back to `self._external_ports`.

**Review by GuillaumeDerval — CHANGES_REQUESTED** [review]

File: base-containers/base/bin/_run_student_intern
Comment (darkyne) [reply base-containers/base/bin/_run_student_intern:65]: Just after getting the user from the grading_container message, adding something like:
`if user=="root" and os.path.exists("/.__input/__shared_kernel"): exit`
would do the trick?

This way even in the case of a compromised grading_container manually sending a message asking to run as root, it would not if student container is using docker... no ?

File: inginious/agent/docker_agent/__init__.py
Comment (darkyne) [reply inginious/agent/docker_agent/__init__.py:451]: Indeed, actually I had in mind cases with a single port needed so the specific case you are talking about was not possible (since "ports" would be empty)... But indeed since the student_container may ask for multiple ports, some could be given to him before realising there are no enough ports available !
Well seen, I would not have remarked this myself.

File: base-containers/base/inginious_container_api/run_student.py
Comment (darkyne) [reply base-containers/base/inginious_container_api/run_student.py:45]: Well actually it should not depend on the runtime.
Well right now the entire "run_student" thing is unavailable for kata indeed....
But when run_student will be available for kata, the "run_as_root" parameter has nothing to do with the docker or kata runtime being selected for the task (in other words, a docker grading_container should be able to start a kata root student_container)

In other words, the entire parameter "run_as_root" is useful for now indeed run_student strats by exiting if run_as_root is set to true but it will have sense in a near future.

File: base-containers/base/inginious_container_api/utils.py
Comment (darkyne) [inline base-containers/base/inginious_container_api/utils.py:66]: From the tests I did, this works well.
It allows the student to play within an interactive shell (launched via the "script" command) and closes the ssh connection when he leaves the shell. The log is written within the ssh session with worker permissions so the student is able to delete the .ssh_logs file (in the task directory) but should probably only code in the "student" directory so the chance of a sad rm * deleting it is low. Only when the student leaves the session, the file is copied into the student directory to be reflected within the grading_container.

**Review by anthonygego — COMMENTED** [review]
"Teardown" is probably a more appropriate and common word than "closure" for this case, which is already used in CS for a different concept.

**Comment by anthonygego** [comment]
LGTM. Waiting for @GuillaumeDerval approval. 
Please do not add any new feature here now.

**Review by anthonygego — APPROVED** [review]

**Comment by GuillaumeDerval** [comment]
LGTM

**Review by GuillaumeDerval — APPROVED** [review]

**Comment by anthonygego** [comment]
As there is one main commit with several fix commits, I'm going to squash everything.

**Pull request merged**
